## Exploring Hacker News Posts

### 1. Import data

In [35]:
from csv import reader
open_file = open('hacker_news.csv')
reader_file = reader(open_file)
hn = list(reader_file)

In [36]:
print(hn[0])
print('\n')
print(hn[1])
print('\n')
print(hn[2])
print('\n')
print(hn[3])
print('\n')
print(hn[4])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


In [37]:
headers = hn[0]
hn = hn[1:]
print(headers)
print('\n')
print(hn[0:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### 2. Extracting Ask HN and Show HN Posts

First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [38]:
ask_posts = []
show_posts = []
other_posts = []

for i in hn:
    if i[1].lower().startswith('ask hn'):
        ask_posts.append(i)
    elif i[1].lower().startswith('show hn'):
        show_posts.append(i)
    else:
        other_posts.append(i)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


### 3. Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [39]:
total_ask_comments = 0
total_show_comments = 0

for i in ask_posts:
    total_ask_comments += int(i[4])

avg_ask_comments = total_ask_comments / len(ask_posts)
print('The average number of comment for ask_posts is ', avg_ask_comments)

for i in show_posts:
    total_show_comments += int(i[4])

avg_show_comments =  total_show_comments / len(show_posts)   
print('The average number of comment for show_posts is ', avg_show_comments)


The average number of comment for ask_posts is  14.038417431192661
The average number of comment for show_posts is  10.31669535283993


On average, ask posts(14) in our sample receive 4 comments more than show posts(10). Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

### 4. Finding the Amount of Ask Posts and Comments by Hour Created

In this part, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

- Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
- Calculate the average number of comments ask posts receive by hour created.

In [40]:
import datetime as dt

In [41]:
'8/4/2016 11:52'

'8/4/2016 11:52'

In [42]:
result_list = []

for i in ask_posts:
    result_list.append([i[6], int(i[4])])
    
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

# i[0] = date, i[1] = number of comments
for i in result_list:
    i[0] = dt.datetime.strptime(i[0], date_format)
    time = i[0].strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = i[1]
    elif time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += i[1]
        
print('The total number of comments for each hour is', comments_by_hour)
print(len(comments_by_hour))


The total number of comments for each hour is {'16': 1814, '22': 479, '15': 4477, '00': 447, '14': 1416, '07': 267, '20': 1722, '11': 641, '04': 337, '12': 687, '17': 1146, '23': 543, '13': 1253, '08': 492, '21': 1745, '18': 1439, '05': 464, '10': 793, '06': 397, '01': 683, '03': 421, '19': 1188, '02': 1381, '09': 251}
24


### 5. Calculating the Average Number of Comments for Ask HN Posts by Hour

In [43]:
avg_by_hour = []

for i in comments_by_hour:
    avg_by_hour.append([i, (comments_by_hour[i] / counts_by_hour[i]) ])
    
avg_by_hour

[['16', 16.796296296296298],
 ['22', 6.746478873239437],
 ['15', 38.5948275862069],
 ['00', 8.127272727272727],
 ['14', 13.233644859813085],
 ['07', 7.852941176470588],
 ['20', 21.525],
 ['11', 11.051724137931034],
 ['04', 7.170212765957447],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['08', 10.25],
 ['21', 16.009174311926607],
 ['18', 13.20183486238532],
 ['05', 10.08695652173913],
 ['10', 13.440677966101696],
 ['06', 9.022727272727273],
 ['01', 11.383333333333333],
 ['03', 7.796296296296297],
 ['19', 10.8],
 ['02', 23.810344827586206],
 ['09', 5.5777777777777775]]

As we can clearly see, 3p.m and 2a.m holds the highest average number of comments

### 6. Sorting and Printing Values from a List of Lists

In [44]:
swap_avg_by_hour = []

for i in avg_by_hour:
    swap_avg_by_hour.append([i[1],i[0]])
    

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [45]:
sorted_swap[:5]

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21']]

In [46]:
print("Top 5 Hours for Ask Posts Comments")
timeformat = "{}: {:.2f} average comments per cost"
for avg, hr in sorted_swap[:5]:
    print(timeformat.format(dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per cost
02:00: 23.81 average comments per cost
20:00: 21.52 average comments per cost
16:00: 16.80 average comments per cost
21:00: 16.01 average comments per cost


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.



### Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.